src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -362,6 +363,22 @@ public void Copy(Operation operation, object objectToApplyTo)
             // Get value at 'from' location and add that value to the 'path' location
             if (TryGetValue(operation.from, objectToApplyTo, operation, out propertyValue))
             {
+                if (propertyValue != null && !propertyValue.GetType().GetTypeInfo().IsValueType)
+                {
+                    // Create deep copy
+                    var copyResult = ConversionResultProvider.ConvertTo(propertyValue, propertyValue.GetType());
+                    if (copyResult.CanBeConverted)
+                    {
+                        propertyValue = copyResult.ConvertedInstance;
+                    }
+                    else
+                    {
+                        var error = CreateOperationFailedError(objectToApplyTo, operation.path, operation, Resources.FormatCannotCopyProperty(operation.from));
+                        ErrorReporter(error);
+                        return;
+                    }
+                }
+
                 Add(operation.path,
```

Comment:
Since Add operation should either do deep copy or maintain reference based on context in which its used, we could add a flag to Add operations of adapters to either maintain reference or not. The reference check could also be moved to `ConversionResultProvider`.